argocd/resource_argocd_application_set.go

```diff
@@ -0,0 +1,322 @@
+			{
+				Severity: diag.Error,
+				Summary: fmt.Sprintf(
+					"progressive sync (`strategy`) is only supported from ArgoCD %s onwards",
+					featureVersionConstraintsMap[featureApplicationSetProgressiveSync].String()),
+				Detail: err.Error(),
+			},
+		}
+	}
+
+	as, err := si.ApplicationSetClient.Create(ctx, &applicationset.ApplicationSetCreateRequest{
+		Applicationset: &application.ApplicationSet{
+			ObjectMeta: objectMeta,
+			Spec:       spec,
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "ApplicationSet",
+				APIVersion: "argoproj.io/v1alpha1",
+			},
+		},
+	})
+	if err != nil {
+		return []diag.Diagnostic{
+			{
+				Severity: diag.Error,
+				Summary:  fmt.Sprintf("application set %s could not be created", objectMeta.Name),
+				Detail:   err.Error(),
+			},
+		}
+	}
+
+	if as == nil {
+		return []diag.Diagnostic{
+			{
+				Severity: diag.Error,
+				Summary:  fmt.Sprintf("application set %s not created: unknown reason", objectMeta.Name),
+			},
+		}
+	}
+
+	d.SetId(as.Name)
+
+	return resourceArgoCDApplicationSetRead(ctx, d, meta)
+}
+
+func resourceArgoCDApplicationSetRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	si := meta.(*ServerInterface)
+	if err := si.initClients(ctx); err != nil {
+		return []diag.Diagnostic{
+			{
+				Severity: diag.Error,
+				Summary:  "failed to init clients",
+				Detail:   err.Error(),
+			},
+		}
+	}
+
+	name := d.Id()
+
+	appSet, err := si.ApplicationSetClient.Get(ctx, &applicationset.ApplicationSetGetQuery{
+		Name: name,
+	})
+	if err != nil {
+		if strings.Contains(err.Error(), "NotFound") {
+			d.SetId("")
+			return diag.Diagnostics{}
+		}
+
+		return []diag.Diagnostic{
+			{
+				Severity: diag.Error,
+				Summary:  "error getting application set",
+				Detail:   err.Error(),
+			},
+		}
+	}
+
+	err = flattenApplicationSet(appSet, d)
+	if err != nil {
+		return []diag.Diagnostic{
+			{
+				Severity: diag.Error,
+				Summary:  fmt.Sprintf("application set %s could not be flattened", name),
+				Detail:   err.Error(),
+			},
+		}
+	}
+
+	return nil
+}
+
+func resourceArgoCDApplicationSetUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	si := meta.(*ServerInterface)
+	if err := si.initClients(ctx); err != nil {
+		return []diag.Diagnostic{
+			{
+				Severity: diag.Error,
+				Summary:  "failed to init clients",
+				Detail:   err.Error(),
+			},
+		}
+	}
+
+	featureApplicationSetSupported, err := si.isFeatureSupported(featureApplicationSet)
+	if err != nil {
+		return []diag.Diagnostic{
+			{
+				Severity: diag.Error,
+				Summary:  "feature not supported",
+				Detail:   err.Error(),
+			},
+		}
+	} else if !featureApplicationSetSupported {
+		return []diag.Diagnostic{
+			{
+				Severity: diag.Error,
+				Summary: fmt.Sprintf(
+					"application set is only supported from ArgoCD %s onwards",
+					featureVersionConstraintsMap[featureApplicationSet].String()),
+			},
+		}
+	}
```

Comment:
Yeah, I think that would be a nice addition.

With regards to the feature "flags" in general, I have been playing around with the following ideas:
- Removing all flags for unsupported versions to clean up the codebase
- Removing the `error` returned by `isFeatureSupported`. The error is ultimately a "developer" error. I.e. passing some random/unknown value to an internal function. So, instead of returning an `error`, we could simply return false (unsupported) for unknown values which would then, assumedly, result in tests failing. Then all of these checks can be simplified down to something like

``` go
if !si.isFeatureSupported(featureApplicationSetSupported) {
	return []diag.Diagnostic{
		{
			Severity: diag.Error,
			Summary: fmt.Sprintf(
				"application set is only supported from ArgoCD %s onwards",
				featureVersionConstraintsMap[featureApplicationSet].String()),
		},
	}
}
```

or even better... combining this with your idea of "an error-to-diag converter" to provide an even terser syntax.

Reply:
Absolutely 👍